File: sdk/go/endpoint.go
package sandbox

import (
	"cmp"
	"crypto/tls"
	"fmt"
	"net/http"
	"net/url"
	"strconv"
	"strings"
)

const httpsScheme = "https"

func (c *Client) configureTLS() error {
	transport := cmp.Or(c.hc.Transport, http.DefaultTransport)
	tr, ok := transport.(*http.Transport)
	if !ok {
		if c.tlsConfig != nil {
			return fmt.Errorf("tls configuration requires an http.Transport")
		}
		return nil
	}
	if c.tlsConfig == nil {
		c.tlsConfig = tr.TLSClientConfig.Clone()
		return nil
	}
	hc := *c.hc
	tr = tr.Clone()
	tr.TLSClientConfig = c.tlsConfig.Clone()
	hc.Transport = tr
	c.hc = &hc
	return nil
}

// WithTLSConfig sets certificate verification for HTTPS requests and agent relays.
func WithTLSConfig(cfg *tls.Config) ClientOption {
	return func(c *Client) { c.tlsConfig = cfg.Clone() }
}

func endpointURL(addr, scheme string) (*url.URL, error) {
	if !strings.Contains(addr, "://") {
		addr = scheme + "://" + addr
	}
	u, err := url.Parse(addr)
	if err != nil {
		return nil, fmt.Errorf("parse sandboxd endpoint: %w", err)
	}
	if (u.Scheme != "http" && u.Scheme != httpsScheme) || u.Hostname() == "" || u.User != nil ||
		(u.Path != "" && u.Path != "/") || strings.ContainsAny(addr, "?#") {
		return nil, fmt.Errorf("sandboxd endpoint must be an http or https origin")
	}
	if port := u.Port(); port != "" {
		n, err := strconv.Atoi(port)
		if err != nil || n < 1 || n > 65535 {
			return nil, fmt.Errorf("sandboxd endpoint port must be between 1 and 65535")
		}
	}
	u.Path = ""
	return u, nil
}

File: sdk/python/cocoonsandbox/endpoint.py
"""HTTP origins shared by control requests and agent relays."""

from __future__ import annotations

import urllib.parse


def _endpoint_url(addr: str, scheme: str = "http") -> urllib.parse.SplitResult:
    if any(ord(c) <= 32 or ord(c) == 127 for c in addr):
        raise ValueError("sandboxd endpoint contains whitespace or a control character")
    if "://" not in addr:
        addr = f"{scheme}://{addr}"
    endpoint = urllib.parse.urlsplit(addr)
    if (
        endpoint.scheme not in ("http", "https")
        or not endpoint.hostname
        or endpoint.username is not None
        or endpoint.path not in ("", "/")
        or "?" in addr
        or "#" in addr
    ):
        raise ValueError("sandboxd endpoint must be an http or https origin")
    try:
        port = endpoint.port
    except ValueError:
        port = 0
    if port == 0:
        raise ValueError("sandboxd endpoint port must be between 1 and 65535")
    return endpoint._replace(path="")
